staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/login.go

```diff
@@ -0,0 +1,160 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"net"
+	"net/http"
+	"strings"
+
+	oidcp "github.com/coreos/go-oidc"
+	"github.com/skratchdot/open-golang/open"
+	"golang.org/x/oauth2"
+)
+
+const (
+	defaultCallbackPort = "8172"
+	defaultCallbackPath = "/authcode"
+
+	successPage = `<!DOCTYPE html>
+<html>
+<head><title>Success</title></head>
+<body onload='window.close()'><p>Login successfully! Please close this browser.</p></body>
+</html>
+`
+)
+
+func (p *oidcAuthProvider) Login() error {
+	issuer := p.cfg[cfgIssuerUrl]
+	clientID := p.cfg[cfgClientID]
+	callbackPort := p.cfg[cfgCallbackPort]
+	if callbackPort == "" {
+		callbackPort = defaultCallbackPort
+	}
+	callbackPath := "/" + strings.TrimLeft(p.cfg[cfgCallbackPath], "/")
+	if callbackPath == "/" {
+		callbackPath = defaultCallbackPath
+	}
+	scopes := []string{oidcp.ScopeOpenID}
+	for _, scope := range strings.Split(p.cfg[cfgExtraScopes], ",") {
+		if scope != "" {
+			scopes = append(scopes, scope)
+		}
+	}
+
+	options := []oauth2.AuthCodeOption{oauth2.AccessTypeOffline}
+	for _, opt := range strings.Split(p.cfg[cfgAuthOptions], ";") {
+		if pos := strings.Index(opt, "="); pos > 0 {
+			options = append(options, oauth2.SetAuthURLParam(opt[:pos], opt[pos+1:]))
+		}
+	}
+
+	ctx := context.WithValue(context.TODO(), oauth2.HTTPClient, p.client)
+	provider, err := oidcp.NewProvider(ctx, issuer)
+	if err != nil {
+		return err
+	}
+	oidcConf := &oidcp.Config{ClientID: clientID}
+	verifier := provider.Verifier(oidcConf)
+	oauth2Conf := oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: p.cfg[cfgClientSecret],
+		Endpoint:     provider.Endpoint(),
+		RedirectURL:  "http://localhost:" + callbackPort + callbackPath,
+		Scopes:       scopes,
+	}
+
+	state, err := generateState()
+	if err != nil {
+		return err
+	}
+
+	shutdownCh := make(chan struct{})
+
+	mux := http.NewServeMux()
+	server := &http.Server{Addr: "127.0.0.1:" + callbackPort, Handler: mux}
+
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, oauth2Conf.AuthCodeURL(state, options...), http.StatusFound)
+	})
+
+	mux.HandleFunc(callbackPath, func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("state") != state {
+			http.Error(w, "invalid session", http.StatusBadRequest)
+			return
+		}
+		token, err := oauth2Conf.Exchange(ctx, r.URL.Query().Get("code"))
+		if err != nil {
+			http.Error(w, "failed to exchange token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		idToken, ok := token.Extra("id_token").(string)
```

Comment:
I would like to say that OIDC should not be required for this. Since `kubectl` can work with `--token`, this should work for any OAuth2 provider (not just the ones who also send id token). How the auth token gets associated with an identity should be up to the `webhook authN server`.

Reply:
Yes. I agree the implementation will be mostly the same. My implementation goes in the OIDC auth plugin in kubectl, so that will be OIDC specific. Regarding OAuth2, it's a different route performed by kubectl. And currently there's no specific design for an interactive flow. I feel we need more discussion on that before we can consolidate the implementation for both flows.